cc/gtp_player.cc

```diff
@@ -0,0 +1,323 @@
+  }
+  return true;
+}
+
+}  // namespace
+
+GtpPlayer::GtpPlayer(std::unique_ptr<DualNet> network, const Options& options)
+    : MctsPlayer(std::move(network), options),
+      num_readouts_(options.num_readouts) {}
+
+bool GtpPlayer::HandleCmd(const std::string& line) {
+  std::vector<absl::string_view> args =
+      absl::StrSplit(line, absl::ByAnyChar(" \t\r\n"), absl::SkipWhitespace());
+  if (args.empty()) {
+    std::cout << "=" << std::endl;
+    return true;
+  }
+
+  auto cmd = args[0];
+  args.erase(args.begin());
+
+  if (cmd == "quit") {
+    std::cout << "=" << std::endl;
+    return false;
+  }
+
+  auto result = DispatchCmd(cmd, args);
+  if (result == "") {
+    std::cout << "=" << std::endl;
+  } else {
+    std::cout << "= " << result << std::endl;
+  }
+  return true;
+}
+
+absl::Span<MctsNode* const> GtpPlayer::TreeSearch(int batch_size) {
+  auto leaves = MctsPlayer::TreeSearch(batch_size);
+  if (!leaves.empty() && report_search_interval_ != absl::ZeroDuration()) {
+    auto now = absl::Now();
+    if (now - last_report_time_ > report_search_interval_) {
+      last_report_time_ = now;
+      ReportSearchStatus(leaves.back());
+    }
+  }
+  return leaves;
+}
+
+void GtpPlayer::ReportSearchStatus(const MctsNode* last_read) {
+  std::cerr << "mg-search:";
+  std::vector<const MctsNode*> path;
+  for (const auto* node = last_read; node != root(); node = node->parent) {
+    path.push_back(node);
+  }
+  for (auto it = path.rbegin(); it != path.rend(); ++it) {
+    std::cerr << " " << (*it)->move.ToKgs();
+  }
+  std::cerr << "\n";
+
+  std::cerr << "mg-q:";
+  for (int i = 0; i < kNumMoves; ++i) {
+    std::cerr << " " << std::fixed << std::setprecision(3)
+              << (root()->child_Q(i) - root()->Q());
+  }
+  std::cerr << "\n";
+
+  std::cerr << "mg-n:";
+  for (const auto& edge : root()->edges) {
+    std::cerr << std::setprecision(0) << " " << edge.N;
+  }
+  std::cerr << "\n";
+
+  std::cerr << "mg-pv:";
+  for (Coord c : root()->MostVisitedPath()) {
+    std::cerr << " " << c.ToKgs();
+  }
+  std::cerr << "\n";
+
+  std::cerr << std::flush;
+}
+
+std::string GtpPlayer::DispatchCmd(absl::string_view cmd,
+                                   const std::vector<absl::string_view>& args) {
+  if (cmd == "clear_board") {
+    return HandleClearBoard(cmd, args);
+  } else if (cmd == "echo") {
+    return HandleEcho(cmd, args);
+  } else if (cmd == "final_score") {
+    return HandleFinalScore(cmd, args);
+  } else if (cmd == "gamestate" || cmd == "mg_gamestate") {
+    return HandleGamestate(cmd, args);
+  } else if (cmd == "genmove" || cmd == "mg_genmove") {
+    return HandleGenmove(cmd, args);
+  } else if (cmd == "play") {
+    return HandlePlay(cmd, args);
+  } else if (cmd == "readouts") {
+    return HandleReadouts(cmd, args);
+  } else if (cmd == "report_search_interval") {
+    return HandleReportSearchInterval(cmd, args);
+  }
+  std::cerr << "ERROR: unrecognized GTP command \"" << cmd << "\"\n";
```

Comment:
http://www.lysator.liu.se/~gunnar/gtp/gtp2-spec-draft2/gtp2-spec.html

2.6 err should start with "?[Id] "

Reply:
done